Clean Data // NLP -> pandas

Pandas - it's used to handle large data with a very efficient way.

pd.read_csv() -> it's used to import a csv formated data.

pd.read_xlsx() -> it's used to import a excel formated data.

pd.read_txt() -> it's used to import a text formated data


In [1]:
import pandas as pd

In [11]:
dataset= pd.read_csv('../Datasets/IPL2025Batters.csv')


In [12]:
dataset.head()# Display the first few rows of the dataset
dataset.head()

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  \
0      Sai Sudharsan   GT   759       15   15   1  108*  54.21  486  156.17   
1  Surya Kumar Yadav   MI   717       16   16   5   73*  65.18  427  167.91   
2        Virat Kohli  RCB   657       15   15   3   73*  54.75  454  144.71   
3       Shubman Gill   GT   650       15   15   2   93*  50.00  417  155.87   
4     Mitchell Marsh  LSG   627       13   13   0   117  48.23  383  163.70   

   100s  50s  4s  6s  
0     1    6  88  21  
1     0    5  69  38  
2     0    8  66  19  
3     0    6  62  24  
4     1    6  56  37

In [13]:
dataset

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  \
0        Sai Sudharsan   GT   759       15   15   1  108*  54.21  486  156.17   
1    Surya Kumar Yadav   MI   717       16   16   5   73*  65.18  427  167.91   
2          Virat Kohli  RCB   657       15   15   3   73*  54.75  454  144.71   
3         Shubman Gill   GT   650       15   15   2   93*  50.00  417  155.87   
4       Mitchell Marsh  LSG   627       13   13   0   117  48.23  383  163.70   
..                 ...  ...   ...      ...  ...  ..   ...    ...  ...     ...   
151      Khaleel Ahmed  CSK     1       14    2   2    1*      -    2   50.00   
152       Mohit Sharma   DC     1        8    2   1    1*   1.00    3   33.33   
153   Donovan Ferreira   DC     1        1    1   0     1   1.00    3   33.33   
154      Digvesh Singh  LSG     1       13    2   1     1   1.00    3   33.33   
155     Manoj Bhandage  RCB     1        1    1   0     1   1.00    4   25.00   

     100s  50s  4s  6s  
0       1    6  88  21  
1       0    5  69  38  
2       0    8  66  19  
3       0    6  62  24  
4       1    6  56  37  
..    ...  ...  ..  ..  
151     0    0   0   0  
152     0    0   0   0  
153     0    0   0   0  
154     0    0   0   0  
155     0    0   0   0  

[156 rows x 14 columns]

In [14]:
dataset.tail(5)  # Display the last 5 rows of the dataset

Player Name Team  Runs  Matches  Inn  No  HS   AVG  BF     SR  100s  \
151     Khaleel Ahmed  CSK     1       14    2   2  1*     -   2  50.00     0   
152      Mohit Sharma   DC     1        8    2   1  1*  1.00   3  33.33     0   
153  Donovan Ferreira   DC     1        1    1   0   1  1.00   3  33.33     0   
154     Digvesh Singh  LSG     1       13    2   1   1  1.00   3  33.33     0   
155    Manoj Bhandage  RCB     1        1    1   0   1  1.00   4  25.00     0   

     50s  4s  6s  
151    0   0   0  
152    0   0   0  
153    0   0   0  
154    0   0   0  
155    0   0   0

In [15]:
# dataset.iloc[start:end:step] its returns rows from start to end with a step
dataset.iloc[75:105:3]

Player Name  Team  Runs  Matches  Inn  No   HS    AVG  BF  \
75         Rahul Tewatia    GT    99       15   12   4  24*  12.38  59   
78         Manish Pandey   KKR    92        3    3   1   37  46.00  65   
81          Marco Jansen  PBKS    75       14    8   4  34*  18.75  63   
84      Romario Shepherd   RCB    70        8    3   1  53*  35.00  24   
87          Jofra Archer    RR    63       12    7   1   30  10.50  50   
90          Harshit Rana   KKR    57       13    7   3   34  14.25  53   
93         Glenn Maxwell  PBKS    48        7    6   0   30   8.00  49   
96           Rashid Khan    GT    40       15    8   3   12   8.00  28   
99   Ravichandran Ashwin   CSK    33        9    4   0   13   8.25  30   
102      Sediqullah Atal    DC    22        1    1   0   22  22.00  16   

         SR  100s  50s  4s  6s  
75   167.79     0    0   6   7  
78   141.53     0    0   5   5  
81   119.04     0    0   3   4  
84   291.66     0    1   5   7  
87   126.00     0    0   3   4  
90   107.54     0    0   5   3  
93    97.95     0    0   5   1  
96   142.85     0    0   2   3  
99   110.00     0    0   3   1  
102  137.50     0    0   0   2

In [16]:
dataset[['Player Name', 'Team', 'HS', 'AVG']].iloc[45:65]

Player Name  Team   HS    AVG
45        Aniket Verma   SRH   74  26.22
46          Will Jacks    MI   53  23.30
47       Dewald Brevis   CSK   57  37.50
48       Hardik Pandya    MI  48*  24.89
49         Nitish Rana    RR   81  21.70
50         Rinku Singh   KKR  38*  29.43
51     Ashutosh Sharma    DC  66*  29.14
52      Faf Du Plessis    DC   62  22.44
53          Karun Nair    DC   89  24.75
54            MS Dhoni   CSK  30*  24.50
55     Rachin Ravindra   CSK  65*  27.29
56           Tim David   RCB  50*  62.33
57  Nitish Kumar Reddy   SRH   32  22.75
58       Shahrukh Khan    GT   57  29.83
59       Andre Russell   KKR  57*  18.56
60         Abdul Samad   LSG   45  20.50
61      Marcus Stoinis  PBKS  44*  26.67
62        Devon Conway   CSK   69  26.00
63        David Miller   LSG  27*  30.60
64     Quinton De Kock   KKR  97*  21.71

In [17]:
#dataset.info() it returns information about the dataset like number of rows, columns, data types, memory usage etc.
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Player Name  156 non-null    object 
 1   Team         156 non-null    object 
 2   Runs         156 non-null    int64  
 3   Matches      156 non-null    int64  
 4   Inn          156 non-null    int64  
 5   No           156 non-null    int64  
 6   HS           156 non-null    object 
 7   AVG          156 non-null    object 
 8   BF           156 non-null    int64  
 9   SR           156 non-null    float64
 10  100s         156 non-null    int64  
 11  50s          156 non-null    int64  
 12  4s           156 non-null    int64  
 13  6s           156 non-null    int64  
dtypes: float64(1), int64(9), object(4)
memory usage: 17.2+ KB


In [18]:
#dataset.describe() it returns statistical summary of the dataset like count, mean, std, min, max, 25%, 50%, 75% etc.
dataset.describe()

Runs     Matches         Inn          No          BF          SR  \
count  156.000000  156.000000  156.000000  156.000000  156.000000  156.000000   
mean   161.320513    9.865385    7.153846    1.634615  105.673077  134.394103   
std    180.352772    4.757455    4.810142    1.658163  112.191838   57.237148   
min      1.000000    1.000000    1.000000    0.000000    1.000000   25.000000   
25%     10.750000    6.000000    2.750000    0.000000   12.000000  100.000000   
50%     93.500000   11.000000    6.000000    1.000000   63.500000  140.005000   
75%    254.250000   14.000000   12.000000    3.000000  164.750000  162.707500   
max    759.000000   17.000000   17.000000    7.000000  486.000000  400.000000   

             100s         50s          4s          6s  
count  156.000000  156.000000  156.000000  156.000000  
mean     0.057692    0.916667   14.391026    8.294872  
std      0.233912    1.590124   17.778774    9.515500  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    1.000000    0.000000  
50%      0.000000    0.000000    7.000000    4.000000  
75%      0.000000    1.000000   23.000000   14.000000  
max      1.000000    8.000000   88.000000   40.000000

In [19]:
dataset.isnull().sum()# Checks for missing (NaN) values in each column and returns the count per column

Player Name    0
Team           0
Runs           0
Matches        0
Inn            0
No             0
HS             0
AVG            0
BF             0
SR             0
100s           0
50s            0
4s             0
6s             0
dtype: int64

In [21]:
students = pd.read_csv("../Datasets/stu_data.csv")

In [22]:
students.head()  # Display the first few rows of the students dataset

Stu_id Stu_name            Stu_email  Stu_age  Stu_hight  Stu_weight
0       1      abc        abc@gmail.com     20.0        5.5        45.0
1       2      efg        efg@gmail.com     24.0        NaN         NaN
2       3      hfd                  NaN     21.0        4.9        54.0
3       4      NaN  something@gmail.com      NaN        5.8         NaN
4       5      gdd        gdd@gmail.com      NaN        NaN        75.0

In [23]:
students.isnull().sum()

Stu_id        0
Stu_name      2
Stu_email     3
Stu_age       4
Stu_hight     3
Stu_weight    4
dtype: int64

In [24]:
students['Stu_name'].fillna("Something", inplace=True)# Fill NaN values in 'Stu_name' column with "Something"   

C:\Users\User\AppData\Local\Temp\ipykernel_15672\3029722293.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  students['Stu_name'].fillna("Something", inplace=True)# Fill NaN values in 'Stu_name' column with "Something"


In [25]:
students

Stu_id   Stu_name            Stu_email  Stu_age  Stu_hight  Stu_weight
0       1        abc        abc@gmail.com     20.0        5.5        45.0
1       2        efg        efg@gmail.com     24.0        NaN         NaN
2       3        hfd                  NaN     21.0        4.9        54.0
3       4  Something  something@gmail.com      NaN        5.8         NaN
4       5        gdd        gdd@gmail.com      NaN        NaN        75.0
5       6  Something                  NaN     48.0        5.9         NaN
6       7        vsc        vsc@gmail.com      NaN        NaN        65.0
7       8        rpm                  NaN     30.0        6.2        58.0
8       9        mno        mno@gmail.com      NaN        5.7         NaN

In [26]:
avg_age = students['Stu_age'].mean()# Calculate the average age of students
students['Stu_age'].fillna(avg_age, inplace=True)  # Fill NaN values in 'Stu_age' column with the average age

C:\Users\User\AppData\Local\Temp\ipykernel_15672\1109385441.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  students['Stu_age'].fillna(avg_age, inplace=True)  # Fill NaN values in 'Stu_age' column with the average age


In [27]:
students

Stu_id   Stu_name            Stu_email  Stu_age  Stu_hight  Stu_weight
0       1        abc        abc@gmail.com     20.0        5.5        45.0
1       2        efg        efg@gmail.com     24.0        NaN         NaN
2       3        hfd                  NaN     21.0        4.9        54.0
3       4  Something  something@gmail.com     28.6        5.8         NaN
4       5        gdd        gdd@gmail.com     28.6        NaN        75.0
5       6  Something                  NaN     48.0        5.9         NaN
6       7        vsc        vsc@gmail.com     28.6        NaN        65.0
7       8        rpm                  NaN     30.0        6.2        58.0
8       9        mno        mno@gmail.com     28.6        5.7         NaN

In [28]:
students.dropna(inplace=True) #  when small number of NaN values are present in the dataset, we can drop those rows

In [29]:
students


Stu_id Stu_name      Stu_email  Stu_age  Stu_hight  Stu_weight
0       1      abc  abc@gmail.com     20.0        5.5        45.0

In [30]:
dataset = pd.read_csv("../Datasets/IPL2025Batters.csv")

In [31]:
dataset

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  \
0        Sai Sudharsan   GT   759       15   15   1  108*  54.21  486  156.17   
1    Surya Kumar Yadav   MI   717       16   16   5   73*  65.18  427  167.91   
2          Virat Kohli  RCB   657       15   15   3   73*  54.75  454  144.71   
3         Shubman Gill   GT   650       15   15   2   93*  50.00  417  155.87   
4       Mitchell Marsh  LSG   627       13   13   0   117  48.23  383  163.70   
..                 ...  ...   ...      ...  ...  ..   ...    ...  ...     ...   
151      Khaleel Ahmed  CSK     1       14    2   2    1*      -    2   50.00   
152       Mohit Sharma   DC     1        8    2   1    1*   1.00    3   33.33   
153   Donovan Ferreira   DC     1        1    1   0     1   1.00    3   33.33   
154      Digvesh Singh  LSG     1       13    2   1     1   1.00    3   33.33   
155     Manoj Bhandage  RCB     1        1    1   0     1   1.00    4   25.00   

     100s  50s  4s  6s  
0       1    6  88  21  
1       0    5  69  38  
2       0    8  66  19  
3       0    6  62  24  
4       1    6  56  37  
..    ...  ...  ..  ..  
151     0    0   0   0  
152     0    0   0   0  
153     0    0   0   0  
154     0    0   0   0  
155     0    0   0   0  

[156 rows x 14 columns]

Feature Extraction/Featurneering -> pe Engiandas

Feature engineering means creating new columns or transforming existing ones to help the model perform better.

In [32]:
def get_runsremark(data):
  if data > 500:
    return "Execellent Run"
  elif data > 300:
    return "Moderate Run"
  elif data > 0:
    return "Poor Run"
  else:
    return "Nothing"

In [33]:
dataset['RunsRemark'] = dataset['Runs'].apply(get_runsremark)  # Creates a new column 'RunsRemark' by applying the function 'get_runsremark' to each value in the 'Runs' column

In [34]:
dataset.head()

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  \
0      Sai Sudharsan   GT   759       15   15   1  108*  54.21  486  156.17   
1  Surya Kumar Yadav   MI   717       16   16   5   73*  65.18  427  167.91   
2        Virat Kohli  RCB   657       15   15   3   73*  54.75  454  144.71   
3       Shubman Gill   GT   650       15   15   2   93*  50.00  417  155.87   
4     Mitchell Marsh  LSG   627       13   13   0   117  48.23  383  163.70   

   100s  50s  4s  6s      RunsRemark  
0     1    6  88  21  Execellent Run  
1     0    5  69  38  Execellent Run  
2     0    8  66  19  Execellent Run  
3     0    6  62  24  Execellent Run  
4     1    6  56  37  Execellent Run

In [35]:
def add_5_runs(data):
  data += 5
  return data

In [ ]:
dataset['Runs'] = dataset['Runs'].apply(add_5_runs)# Adds 5 to each value in the 'Runs' column using the 'add_5_runs' function

In [37]:
dataset.head()

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  \
0      Sai Sudharsan   GT   764       15   15   1  108*  54.21  486  156.17   
1  Surya Kumar Yadav   MI   722       16   16   5   73*  65.18  427  167.91   
2        Virat Kohli  RCB   662       15   15   3   73*  54.75  454  144.71   
3       Shubman Gill   GT   655       15   15   2   93*  50.00  417  155.87   
4     Mitchell Marsh  LSG   632       13   13   0   117  48.23  383  163.70   

   100s  50s  4s  6s      RunsRemark  
0     1    6  88  21  Execellent Run  
1     0    5  69  38  Execellent Run  
2     0    8  66  19  Execellent Run  
3     0    6  62  24  Execellent Run  
4     1    6  56  37  Execellent Run

unique()

In [38]:
dataset['Team'].unique() # Returns an array of all unique values (team names) in the 'Team' column

array(['GT', 'MI', 'RCB', 'LSG', 'PBKS', 'RR', 'DC', 'SRH', 'KKR', 'CSK'],
      dtype=object)

In [39]:
len(dataset['Team'].unique())

10

In [40]:
len(dataset[dataset['Team'] == "GT"]) # Returns the number of rows where the 'Team' column is equal to "GT"


14

In [41]:
dataset[(dataset['Team'] == "GT") & (dataset['RunsRemark'] == "Execellent Run")]# Filters and returns rows where the team is "GT" AND the 'RunsRemark' is "Execellent Run"

Player Name Team  Runs  Matches  Inn  No    HS    AVG   BF      SR  100s  \
0  Sai Sudharsan   GT   764       15   15   1  108*  54.21  486  156.17     1   
3   Shubman Gill   GT   655       15   15   2   93*  50.00  417  155.87     0   
9    Jos Buttler   GT   543       14   13   4   97*  59.78  330  163.03     0   

   50s  4s  6s      RunsRemark  
0    6  88  21  Execellent Run  
3    6  62  24  Execellent Run  
9    5  52  24  Execellent Run

value_counts() 

In [42]:
dataset['Team'].value_counts() # Returns the count of occurrences of each unique value in the 'Team' column, sorted in descending order

Team
CSK     19
RR      17
DC      17
MI      16
KKR     15
LSG     15
PBKS    15
GT      14
RCB     14
SRH     14
Name: count, dtype: int64

In [43]:
dataset['RunsRemark'].value_counts()

RunsRemark
Poor Run          125
Moderate Run       20
Execellent Run     11
Name: count, dtype: int64

groupby() is used to:

Aggregate values (mean, sum, count, etc.) for each group.

Generate new features based on group-level statistics.

Understand patterns in subsets of data (like per user, per product, per team, etc.)

In [44]:
dataset.groupby('Team')['Runs'].max()

Team
CSK     362
DC      544
GT      764
KKR     395
LSG     632
MI      722
PBKS    609
RCB     662
RR      564
SRH     492
Name: Runs, dtype: int64

In [50]:
avg_runs = dataset.groupby('Team')['Runs'].mean()

In [51]:
avg_runs

Team
CSK     126.842105
DC      145.352941
GT      202.571429
KKR     130.733333
LSG     178.200000
MI      180.125000
PBKS    205.000000
RCB     186.357143
RR      151.823529
SRH     174.857143
Name: Runs, dtype: float64

In [52]:
avg_runs = dataset.groupby('Player Name')['Runs'].mean()

In [76]:
dataset.groupby('Team')['Runs'].mean()

Team
CSK     126.842105
DC      145.352941
GT      202.571429
KKR     130.733333
LSG     178.200000
MI      180.125000
PBKS    205.000000
RCB     186.357143
RR      151.823529
SRH     174.857143
Name: Runs, dtype: float64

In [53]:
avg_runs

Player Name
Abdul Samad         169.0
Abhinav Manohar      66.0
Abhishek Sharma     444.0
Abishek Porel       306.0
Aiden Markram       450.0
                    ...  
Wiaan Mulder         14.0
Will Jacks          238.0
Xavier Bartlett      16.0
Yash Dayal            9.0
Yashasvi Jaiswal    564.0
Name: Runs, Length: 156, dtype: float64

reset_index() is used to:

🔄 Convert the index of a DataFrame (or Series) into normal columns and reset the index to default (0, 1, 2,...).

In [63]:
avg = dataset.groupby('Team')['Runs'].mean().reset_index()

In [64]:
avg

Team        Runs
0   CSK  126.842105
1    DC  145.352941
2    GT  202.571429
3   KKR  130.733333
4   LSG  178.200000
5    MI  180.125000
6  PBKS  205.000000
7   RCB  186.357143
8    RR  151.823529
9   SRH  174.857143

In [60]:
dataset.groupby('Team')['Runs'].max().sort_values(ascending=False)# Groups the dataset by 'Team', finds the maximum 'Runs' scored by each team, and sorts the result in descending order
# dataset.groupby('Team')
# → Groups the rows based on each unique team in the 'Team' column.

# ['Runs'].max()
# → For each team group, it finds the maximum value in the 'Runs' column (i.e., the highest run scored by a player in that team).

# .sort_values(ascending=False)
# → Sorts the resulting series in descending order, so the team with the highest individual run appears first.

Team
GT      764
MI      722
RCB     662
LSG     632
PBKS    609
RR      564
DC      544
SRH     492
KKR     395
CSK     362
Name: Runs, dtype: int64


🔍 loc in Pandas — Explained Simply
loc stands for "location". It is used to access rows and columns in a Pandas DataFrame by label (not by position).

> df.loc[row_label, column_label]


In [ ]:
#Access a row by label:
dataset.loc[2]   # Gets the row where index == 2


Player Name       Virat Kohli
Team                      RCB
Runs                      662
Matches                    15
Inn                        15
No                          3
HS                        73*
AVG                     54.75
BF                        454
SR                     144.71
100s                        0
50s                         8
4s                         66
6s                         19
RunsRemark     Execellent Run
Name: 2, dtype: object

In [66]:
# Access a specific value:
dataset.loc[2, 'Player Name']   # Value at row index 2, column "Player Name"


'Virat Kohli'

In [68]:
# Access multiple columns for a specific row:
dataset.loc[2, ['Player Name', 'Runs']]


Player Name    Virat Kohli
Runs                   662
Name: 2, dtype: object

In [70]:
dataset.loc[dataset['Team'] == 'RCB']

Player Name Team  Runs  Matches  Inn  No   HS    AVG   BF      SR  \
2          Virat Kohli  RCB   662       15   15   3  73*  54.75  454  144.71   
16           Phil Salt  RCB   408       13   13   1   65  33.58  229  175.98   
28       Rajat Patidar  RCB   317       15   14   1   64  24.00  217  143.77   
38       Jitesh Sharma  RCB   266       15   11   4  85*  37.29  148  176.35   
41    Devdutt Padikkal  RCB   252       10   10   1   61  27.44  164  150.60   
56           Tim David  RCB   192       12    9   6  50*  62.33  101  185.14   
72    Liam Livingstone  RCB   117       10    8   1   54  16.00   84  133.33   
73       Krunal Pandya  RCB   114       15    7   1  73*  18.17   86  126.74   
77      Mayank Agarwal  RCB   100        4    4   1  41*  31.67   64  148.43   
84    Romario Shepherd  RCB    75        8    3   1  53*  35.00   24  291.66   
86       Jacob Bethell  RCB    72        2    2   0   55  33.50   39  171.79   
112  Bhuvneshwar Kumar  RCB    19       14    6   3    8   4.67   25   56.00   
138         Yash Dayal  RCB     9       15    3   1    3   2.00    8   50.00   
155     Manoj Bhandage  RCB     6        1    1   0    1   1.00    4   25.00   

     100s  50s  4s  6s      RunsRemark  
2       0    8  66  19  Execellent Run  
16      0    4  48  22    Moderate Run  
28      0    2  25  14    Moderate Run  
38      0    1  24  17        Poor Run  
41      0    2  21  14        Poor Run  
56      0    1  16  14        Poor Run  
72      0    1   4   9        Poor Run  
73      0    1   9   4        Poor Run  
77      0    0  10   2        Poor Run  
84      0    1   5   7        Poor Run  
86      0    1   9   3        Poor Run  
112     0    0   1   0        Poor Run  
138     0    0   0   0        Poor Run  
155     0    0   0   0        Poor Run

In [72]:
dataset.loc[dataset['Runs'] > 500, 'Player Name']
# Access a specific row and column based on a condition:

0         Sai Sudharsan
1     Surya Kumar Yadav
2           Virat Kohli
3          Shubman Gill
4        Mitchell Marsh
5          Shreyas Iyer
6      Yashasvi Jaiswal
7     Prabhsimran Singh
8             K L Rahul
9           Jos Buttler
10      Nicholas Pooran
Name: Player Name, dtype: object

🔍 iloc in Pandas — Explained Simply
iloc stands for "integer location". It is used to access rows and columns by position/index number, not by label.

✅ Syntax:
df.iloc[row_index, column_index]

In [73]:
#. Access a specific row (by number)
dataset.iloc[0]    # First row


Player Name     Sai Sudharsan
Team                       GT
Runs                      764
Matches                    15
Inn                        15
No                          1
HS                       108*
AVG                     54.21
BF                        486
SR                     156.17
100s                        1
50s                         6
4s                         88
6s                         21
RunsRemark     Execellent Run
Name: 0, dtype: object

In [74]:
dataset.iloc[2, 1]   # Row 2, Column 1 (3rd row, 2nd column)


'RCB'

In [75]:

dataset.iloc[0, [1, 2]]   # First row, 2nd and 3rd columns

Team     GT
Runs    764
Name: 0, dtype: object

| Feature  | `loc`                              | `iloc`                              |
| -------- | ---------------------------------- | ----------------------------------- |
| Uses     | Labels (column names, index names) | Integer index positions (0,1,2,...) |
| Example  | `df.loc[2, 'Runs']`                | `df.iloc[2, 1]`                     |
| Safe for | Named access, filtering            | Positional access                   |
---------------------------------------------------------------------------------------

 Find player names along with their maximum runs for each team as  best players.

In [77]:
if 'Runs' in dataset.columns:
    df_to_analyze = dataset
    stat_column = 'Runs'
    best_players = df_to_analyze.loc[df_to_analyze.groupby('Team')[stat_column].idxmax()][['Player Name', 'Team', stat_column]]
    best_players = best_players.rename(columns={stat_column: 'Stat'})

else:
    raise ValueError("CSV must contain either 'Runs'")

In [79]:
best_players

Player Name  Team  Stat
22        Shivam Dube   CSK   362
8           K L Rahul    DC   544
0       Sai Sudharsan    GT   764
18     Ajinkya Rahane   KKR   395
4      Mitchell Marsh   LSG   632
1   Surya Kumar Yadav    MI   722
5        Shreyas Iyer  PBKS   609
2         Virat Kohli   RCB   662
6    Yashasvi Jaiswal    RR   564
11   Heinrich Klaasen   SRH   492

.loc[] is used to access rows and columns in a DataFrame by label (not position).
General Syntax:

df.loc[row_label, column_label]
row_label: index (or list of indices) you want to select.

column_label: column name (or list of names) to return.
--------------------------------------------------------------
# Get the index of the row where each team's player has the maximum Runs
max_run_indices = dataset.groupby('Team')['Runs'].idxmax()

# Use those indices to get the corresponding player names, team, and runs
top_scorers = dataset.loc[max_run_indices, ['Player Name', 'Team', 'Runs']]

# Sort by Runs in descending order
top_scorers = top_scorers.sort_values(by='Runs', ascending=False)


🧠 Explanation:
groupby('Team')['Runs'].idxmax()
→ Finds the index of the player with the highest runs in each team.

dataset.loc[...]
→ Extracts full row details (player name, team, runs).

sort_values(by='Runs', ascending=False)
→ Ranks all top scorers from highest to lowest.
===================================================

 Dataset is:

index	Player Name	Team	Runs
0	    Kohli	    RCB	  145
1	    Rohit     MI	  120
2	    Ruturaj 	CSK	  132
3	    Gill	    GT	  115

Then:


max_run_indices = [0, 1, 2, 3]
dataset.loc[max_run_indices, ['Player Name', 'Team', 'Runs']]

Returns the top run-scorer from each team, as a filtered table.

In [78]:
dataset.loc[dataset.groupby('Team')['Runs'].idxmax(), ['Player Name', 'Team', 'Runs']].sort_values(by='Runs', ascending=False)

# What this does:
# groupby('Team')['Runs'].idxmax() → Gets the index of the row with max runs per team

# loc[...] → Selects those rows from the DataFrame

# [['Player Name', 'Team', 'Runs']] → Picks relevant columns

# .sort_values(by='Runs', ascending=False) → Sorts by highest runs

#-----------------------------------------------------------------------

# ✅ What .idxmax() means:
# .idxmax() returns the index (row label) of the maximum value in a Series or within each group in a groupby().

# dataset.groupby('Team')['Runs'].idxmax()
# This returns, for each team, the row index of the player who has the highest runs.

#----------------------------------------------------------------------------
# dataset.groupby('Team')['Runs'].idxmax()
# This returns the index of the row where each team's player scored the maximum runs.

# dataset.loc[...]
# # Now we use .loc[...] to fetch the full data from those rows.
# Means:

# Fetch only the rows with max-run players

# And from those rows, return only the 'Player Name', 'Team', and 'Runs' columns




Player Name  Team  Runs
0       Sai Sudharsan    GT   764
1   Surya Kumar Yadav    MI   722
2         Virat Kohli   RCB   662
4      Mitchell Marsh   LSG   632
5        Shreyas Iyer  PBKS   609
6    Yashasvi Jaiswal    RR   564
8           K L Rahul    DC   544
11   Heinrich Klaasen   SRH   492
18     Ajinkya Rahane   KKR   395
22        Shivam Dube   CSK   362

In [ ]:
dataset.groupby('Team')['Runs'].idxmax()
# This line returns the row indices (from the original DataFrame) where each team’s player scored the maximum runs.

Team
CSK     22
DC       8
GT       0
KKR     18
LSG      4
MI       1
PBKS     5
RCB      2
RR       6
SRH     11
Name: Runs, dtype: int64